File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/operators/impl/BitSetOnePointCrossover.java
package ar.edu.ungs.yamiko.ga.operators.impl;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;

import ar.edu.ungs.yamiko.ga.domain.Individual;
import ar.edu.ungs.yamiko.ga.exceptions.NullIndividualException;
import ar.edu.ungs.yamiko.ga.exceptions.YamikoException;
import ar.edu.ungs.yamiko.ga.operators.Crossover;
import ar.edu.ungs.yamiko.ga.toolkit.BitsStaticHelper;
import ar.edu.ungs.yamiko.ga.toolkit.StaticHelper;

/**
 * Operador de Crossover en un punto implementado para individuos basados en tiras de bits.
 * FIXME: Funciona para individuos con un cromosoma solo.
 * 
 * @author ricardo
 *
 */
public class BitSetOnePointCrossover implements Crossover<BitSet>{

	public BitSetOnePointCrossover() {

	}
	
	public List<Individual<BitSet>> execute(List<Individual<BitSet>> individuals) throws YamikoException {
		if (individuals==null) throw new NullIndividualException();
		if (individuals.size()<2) throw new NullIndividualException();
		if (individuals.get(0)==null || individuals.get(1)==null) throw new NullIndividualException();
		List<Individual<BitSet>> descendants=new ArrayList<Individual<BitSet>>();
		Individual<BitSet> i1 = individuals.get(0);
		Individual<BitSet> i2 = individuals.get(1);		

		// FIXME Solo 1 cromosoma
//		int i=0;
//		for (Chromosome<BitSet> c: i1.getGenotype().getChromosomes()) {
//				BitSet bs1=c.getFullRawRepresentation();
//				int point=StaticHelper.randomInt(bs1.length());
//				bs1.0[2];
//				BitSet 
//				i++;
//		}
		
		BitSet c1=i1.getGenotype().getChromosomes().iterator().next().getFullRawRepresentation();
		BitSet c2=i2.getGenotype().getChromosomes().iterator().next().getFullRawRepresentation();
		int point=StaticHelper.randomInt(c1.size());
		BitSet desc1=new BitSet(c1.length());
		BitSet desc2=new BitSet(c1.length());
		for (int i=0;i<c1.size();i++)
		{
			if (i<point)
			{
				desc1.set(i,c1.get(i));
				desc2.set(i,c2.get(i));
			}
			else
			{
				desc1.set(i,c2.get(i));
				desc2.set(i,c1.get(i));				
			}
		}
		
		Individual<BitSet> d1=BitsStaticHelper.create(i1.getGenotype().getChromosomes().get(0).name(), desc1);
		Individual<BitSet> d2=BitsStaticHelper.create(i2.getGenotype().getChromosomes().get(0).name(), desc2);		
		descendants.add(d1);
		descendants.add(d2);
		return descendants;
		
	}
}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/operators/impl/BitSetOnePointCrossover.scala
package ar.edu.ungs.yamiko.ga.operators.impl

import ar.edu.ungs.yamiko.ga.operators.Crossover
import ar.edu.ungs.yamiko.ga.domain.Individual
import ar.edu.ungs.yamiko.ga.exceptions.NullIndividualException
import scala.util.Random
import ar.edu.ungs.yamiko.ga.toolkit.IndividualBitSetFactory
import scala.collection.mutable.BitSet


/**
 * Operador de Crossover en un punto implementado para individuos basados en tiras de bits.
 * FIXME: Funciona para individuos con un cromosoma solo.
 * 
 * @author ricardo
 *
 */
@SerialVersionUID(31119L)
class BitSetOnePointCrossover extends Crossover[BitSet] {
     
    override def execute(individuals:List[Individual[BitSet]]):List[Individual[BitSet]] = {

		  if (individuals==null) throw new NullIndividualException("BitSetOnePointCrossover")
		  if (individuals.length<2) throw new NullIndividualException("BitSetOnePointCrossover")
		  if (individuals(0)==null || individuals(1)==null) throw new NullIndividualException("BitSetOnePointCrossover");
		
		  val i1 = individuals(0)
		  val i2 = individuals(1)
		
		  val c1:BitSet=i1.getGenotype().getChromosomes()(0).getFullRawRepresentation()
		  val c2:BitSet=i2.getGenotype().getChromosomes()(0).getFullRawRepresentation()
		
		  val r = Random
      val point=r.nextInt(c1.size)
		
      val desc1=new scala.collection.mutable.BitSet(c1.size)
		  val desc2=new scala.collection.mutable.BitSet(c1.size)
		
		  for (i <- 0 to c1.size-1)
			  if (i<point)
			  {
			    if (c1(i)) desc1.add(i)
			    if (c2(i)) desc2.add(i)
			  }
			  else
  			{
			    if (c1(i)) desc2.add(i)
			    if (c2(i)) desc1.add(i)
  			}
		
		    val d1:Individual[BitSet]= IndividualBitSetFactory.create(i1.getGenotype().getChromosomes()(0).name(), desc1)
		    val d2:Individual[BitSet]= IndividualBitSetFactory.create(i2.getGenotype().getChromosomes()(0).name(), desc2)
		
		  return List(d1,d2)		      
      
     }

}